### PR TITLE
Add disaggregated prefill impl for gemma4-31b-it

### PR DESCRIPTION
**Pull request by svuckovicTT**
### PR Category
Feature

### Summary
Add standalone Gemma4 prefill support for disaggregated inference on a single Galaxy. Supports CP8/TP4 and CP4/TP8 layouts with chunked, context-parallel ring attention and packed KV caches.

### Notes for reviewers
Implementation lives in `models/demos/gemma4_d_p` and includes KV-cache migration helpers and traced full-model and layer benchmarks. This adds the prefill component; decode workers and end-to-end serving orchestration are outside its scope.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:svuckovic/gemma4-prefill-model)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:svuckovic/gemma4-prefill-model)

### CODE REVIEW

**Review by dpopovTT — APPROVED** [review]

File: models/demos/gemma4_d_p/tt/attention/prefill.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        kv_replicated=weights.kv_replicated,
+        memory_config=act_mc,
+        kv_tied=kv_tied,
+    )
+
+    tt_q = apply_per_head_norm(tt_q, weights.q_norm_weight, config.rms_norm_eps, with_scale=True, memory_config=act_mc)
+
+    packed_global_ring = weights.is_global and isinstance(ring_kv_cache, PackedRingKVCache)
+    packed_sliding_ring = config.is_sliding and ring_kv_cache is not None
+    if weights.is_global and kv_tied:
```
Comment (yieldthought) [inline models/demos/gemma4_d_p/tt/attention/prefill.py]: Codex: **[P1] New/different — Keep the same tied-KV normalization when the cache falls back to a projection with duplicate V columns.** Both projections represent the same semantic K=V vector. The current full-QKV branch sends packed, unrotated K to the canonical-K path. This condition change makes the host numerical probe agree with the tied-QK path.

Suggestion:
```suggestion
    if weights.is_global:
```

File: models/demos/gemma4_d_p/tt/common.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    # variant, a marker seeded under the old overrides would certify a warm build whose files do
+    # not exist -- and as_tensor would persist placeholders for them. (#45400 review, finding B2)
+    _precision_for_variant = Gemma4Precision.load(model_path, _worker_mesh)
+    cache_identity = dict(
+        model_name=os.path.basename(str(model_path).rstrip("/")) or "gemma4",
+        n_layers=model_args.num_hidden_layers,
+        mesh_shape=_worker_mesh,
+        build_variant={
+            "precision": {k: str(v) for k, v in sorted(_precision_for_variant._overrides.items())},
+        },
```
Comment (yieldthought) [inline models/demos/gemma4_d_p/tt/common.py:100]: Codex: **[P1] New/different — Invalidate warm markers when the required projection cache changes.** Switching `GEMMA4_TIED_QKV` from `0` to `1` currently accepts the old marker and constructs the missing QK cache from placeholders. Include the effective option and a layout revision so existing unversioned markers require real host weights. The host marker probe verifies that this variant rejects the old marker.

Suggestion:
```suggestion
        build_variant={
            "prefill_cache_layout": 1,
            "tied_qkv": os.environ.get("GEMMA4_TIED_QKV", "1") == "1",
            "precision": {k: str(v) for k, v in sorted(_precision_for_variant._overrides.items())},
        },
```

**Comment by yieldthought** [comment]
Codex review:

1. **[P1] New/different — Preserve tied-KV semantics in the legacy QKV fallback.** [models/demos/gemma4_d_p/tt/attention/prefill.py:71](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/tt/attention/prefill.py#L71)

When the cache has `wqkv_packed640` but no `wqk_packed640`, `load_attention_weights` automatically selects the full QKV projection. This branch then normalizes K with the unpermuted gamma, leaves K unrotated, and passes it as `canonical_k` to `pack_global_kv_device`, even though its projection is in packed V order. The packed helper gathers canonical rotary indices from those wrong channels. A host execution of the actual loader/prefill functions with CPU operation stand-ins gives output PCC 0.528 against the tied-QK path; V agrees exactly, but K's rotary prefix differs by up to 12.02. Suggestion: apply the tied-KV normalization/packing path whenever `weights.is_global`, independently of whether the projection includes duplicate V columns. The one-line inline suggestion restores equality in this probe.

2. **[P1] New/different — Include the projection choice in the warm-cache identity.** [models/demos/gemma4_d_p/tt/common.py:95](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/tt/common.py#L95)

`GEMMA4_TIED_QKV` changes the required filename from `wqkv_packed640` to `wqk_packed640`, but `build_variant` records only precision. A complete cache built with the flag set to `0` therefore passes the warm gate after switching to the default `1`. `CachedStateDict` supplies placeholder tensors, making `qk` non-None and bypassing the missing-QK fallback in [models/demos/gemma4_d_p/tt/attention/weights.py:224](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/tt/attention/weights.py#L224). `as_tensor` then creates the missing QK cache from uninitialized placeholder data. The host probe confirms the old marker is accepted and the new QK filename receives a placeholder-derived tensor. Suggestion: version the prefill cache layout and include the effective tied-QKV option in `build_variant`, so this transition reloads real weights.

3. **[P2] New/different — Handle a single chunk before entering ring sliding attention.** [models/demos/gemma4_d_p/tt/common.py:43](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/tt/common.py#L43)

The defaults accept `max_seq_len=8192` and `prefill_chunk_size=8192`. Each rank then has equal Q and cache sequence extents (1024 on CP8, 2048 on CP4), so `RingJointSDPAInputs::is_chunked()` is false. The sliding-window operation explicitly rejects that case with `Ring sliding-window attention requires chunked prefill` ([ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:594](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp#L594)). The parametrized `ctx_32k/chunk32768` demo cases have the same problem. Suggestion: provide a supported single-chunk path or larger physical cache capacity, and test the default constructor plus one-chunk prefill on both layouts.

4. **[P2] Duplicate — Add a numerical reference check for full-model replay.** [models/demos/gemma4_d_p/demo/text_demo_prefill.py:405](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/demo/text_demo_prefill.py#L405)

The current full-model test still checks only finiteness; there is no reference-output comparison covering layer composition and a later chunk. This remains the issue raised in [the earlier Copilot comment](https://github.com/tenstorrent/tt-metal/pull/55748#discussion_r3958460845). Suggestion: compare against the reference with a stated numerical threshold and exercise both initial and later-chunk replay, including the legacy projection path.

5. **[P2] Duplicate — Register this package in Models CI.** [models/demos/gemma4_d_p/demo/text_demo_prefill.py:264](https://github.com/tenstorrent/tt-metal/blob/d3064a5fd6b03b36bda2d2d5d77067311b39f3bb/models/demos/gemma4_d_p/demo/text_demo_prefill.py#L264)

The package remains absent from the tiered unit/e2e registries, model tiers/targets, workflow model enums, and time budget. The green PR checks therefore do not establish coverage of these new tests. This repeats [the earlier CI-registration comment](https://github.com/tenstorrent/tt-metal/pull/55748#discussion_r3958460797). Suggestion: add the required model registration and measured test budgets.

Validation: manual review at d3064a5fd6b03b36bda2d2d5d77067311b39f3bb; focused host numerical/cache probes and static API tracing. All 33 Python files parse and the diff passes whitespace checks. No Galaxy execution or performance validation was performed; the normal pytest suite cannot collect on this host because compiled TTNN is unavailable.

**Comment by svuckovicTT** [comment]
@yieldthought addressed the suggestions. CI sanity and accuracy testing is actively being worked on in parallel by other folks, should land by EoW following this PR.

File: models/demos/gemma4_d_p/demo/text_demo_prefill.py [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+# ── Traced long-context chunked prefill (production shape) ────────────────────
+
+
+@torch.no_grad()
+@parametrize_mesh_with_fabric([(8, 4), (4, 8)], device_params_extra={"trace_region_size": TRACE_REGION_SIZE})
+@pytest.mark.parametrize("token_source", ["text"], ids=lambda t: t)
+@pytest.mark.parametrize("chunk_size", PREFILL_CHUNK_SIZES, ids=lambda c: f"chunk{c}")
+@pytest.mark.parametrize("context_len", [32768, 65536, 131072, 262144], ids=lambda c: f"ctx_{c // 1024}k")
+@pytest.mark.parametrize("readback_all", [True, False], ids=["readback_all", "readback_final"])
+def test_prefill_long_context_traced(
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/demo/text_demo_prefill.py:214]: Will be done in a follow-up.

File: models/demos/gemma4_d_p/demo/text_demo_prefill.py [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+            # Reading every chunk's hidden states to host is a test artifact — a prefill
+            # server leaves the KV cache on device and reads back only the last chunk,
+            # whose final row seeds the first decode step. readback="final" measures that
+            # shape; "all" gathers every chunk, which costs wall time but checks each one
+            # for finiteness instead of only the last.
+            if readback_all or chunk_idx == n_chunks - 1:
+                t_rb = time.time()
+                hidden = _cp_gather_torch(out, mesh_device, mesh_config)
+                assert torch.isfinite(hidden).all(), f"chunk {chunk_idx} produced non-finite output"
+                readback_s += time.time() - t_rb
```
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply models/demos/gemma4_d_p/demo/text_demo_prefill.py:356]: Will be done in a follow-up.

**Comment by yieldthought** [comment]
Codex review: Re-reviewed at `bf26dc5ef7d5ea538238a439f7be711b6bbd4c38`.

The three implementation findings are resolved: global attention now always uses tied QK, the new cache identity rejects old completion markers, and physical KV capacity keeps single-chunk prefill on the supported ring path. Focused host checks passed; I found no new correctness defects in this follow-up.

Two previously raised validation items remain open, as noted in [your follow-up](https://github.com/tenstorrent/tt-metal/pull/55748#issuecomment-5599660503):

- **[P2] Full-model numerical validation:** the traced test still [checks only finiteness](https://github.com/tenstorrent/tt-metal/blob/bf26dc5ef7d5ea538238a439f7be711b6bbd4c38/models/demos/gemma4_d_p/demo/text_demo_prefill.py#L355). Suggestion: land a reference-output comparison with a stated threshold covering the first and a later chunk.
- **[P2] CI registration:** `gemma4_d_p` is still absent from the tiered unit/e2e registries and model tiers/targets. Suggestion: register the package and run its validation tests so the green PR checks cover the new model path.

The planned follow-up does not yet provide that coverage, so these two items remain unresolved. Verification here was host/source only; normal pytest collection requires compiled TTNN, and no Galaxy run was performed.

**Review by yieldthought — APPROVED** [review]
Approved to unblock, I think tests are important though
